### ISSUE TITLE
Synchronisation drift/delay with Brainproducts / LabRecorder 

### ISSUE THREAD

**Issue by ioannis-dz** [sccn/labstreaminglayer#43]
I am using the following set up:

* Recording acceleration data and EEG using the brainamp EEG/EXG amps and the acceleration sensor.
* Recording collusion events from a servo in the Unity engine, using the following code:
```CSharp
// Setup the stream when starting
 markinf = new liblsl.StreamInfo("UnityMarkers", "Markers", 1, 0, liblsl.channel_format_t.cf_string, "UnityMarkerReadings");
 markoutl = new liblsl.StreamOutlet(markinf);

// Code that runs every 4ms to talk to the servo
 // if hit event detected
   // Sent event to LSL
    markoutl.push_sample(EventMarkers, liblsl.local_clock());
```
* Saving the streams in LabRecorder 1.10
* Importing the data in EEGLAB with xdfimport 1.12 

The servo is set to hit the accelerometer periodically to see that the two streams are synchronised (and the delay of detecting a hit). All the recording is done using a single computer (Win 7).

The curious problem is that the hit event recorded in Unity (marked by the red Hit event) is always faster than the accelerometer data recorded by the EXG (blue line). This drift depends on the method that the brainamp is being recorded:

i. a 30 ms (approx.) drift as shown below  when recording with the BrainAmpSeries exe.(Apps 1.10)  ![brainampseries - 30 ms drift](https://cloud.githubusercontent.com/assets/12684882/7903102/b87eef58-07c9-11e5-879d-12870c97cdea.png)

ii. a 40 ms (approx,) drift as shown below when recording with the BrainVisionRDA exe (Apps 1.10) ![brainvisionrda - 40 ms drift](https://cloud.githubusercontent.com/assets/12684882/7903103/bf71c448-07c9-11e5-9331-2c71123d68f9.png) 

iii. a 50 ms (approx,) drift as shown below when recording with OVAS 0.14.3 and the brainvision Recorder in remote access. 
![ovas from brainvision recorder 50 ms delay](https://cloud.githubusercontent.com/assets/12684882/7903109/e4955262-07c9-11e5-9403-3e034dd9be09.png) 

I have measured (using an oscilloscope) that the time Unity takes to register the hit event is roughly 30 ms (+/- 5 ms). I have come up with two potential reasons for this drift/delay in the two stream: 

1. There is an approx. 50 ms - 90 ms latency (30 ms delay to detect the event in Unity plus the extra delay shown above) in recording brainproduct data depending on how they are recorded (BrainAmpseries,exe, etc.)
2. When imported by the xdfimporter the clocks from the two streams (brainamp and unityevents) are not properly aligned.

In case of 1) I haven't found any evidence to support such a high latency from the brainamp hardware so is this latency coming from the recording software (in Apps)? 

Is there a mistake in my method, or anything I have missed? Is this high latency to be expected?

Thank you very much in advance.

**Comment by dmedine** [comment]
ioannis,
It is unlikely that anything is not properly aligned at the importing stage. There is also some amount of software latency depending on you computer and OS etc. These latencies you report are not terribly outlandish given other measurements that I have seen and taken.

Having said that, 30-50ms in the eeg recording data does seem like a lot. One question, are the latencies consistent from method to method? E.g, does BrainVisionRDA.exe always come ~40ms late? 

I have done a number of synchronization/latency experiments already, although not with the brainamp products. You can read about it here:
http://sccn.ucsd.edu/labinfo/SCCN_latencies/index.php

I will attempt to make my own measurements if I can find some brainamp equipment lying around somewhere. 

Also, if the latencies are consistent, you can easily compensate for synchronization purposes. You can add a line to the xdf header (using the XDFBowser app) such as this one:
<desc>
  <synchronization>        
    <offset_mean>0.00772</offset_mean>    
  </synchronization>
</desc>
This would add 7.72ms to every timestamp in this particular data stream.

**Comment by ioannis-dz** [comment]
David, 

Thank you very much for your response. 

A few comments based on your reply:

1. I have run a couple of tests to identify the stability of the delay, depending on the method the range is between 10 - 20 ms. (Eg. BrainAmp.exe has a range of 10 ms from 23-33 ms)

2. The actual delay is much larger. The hit event is already delayed by about 30 ms so the actual delay can be up to almost 100 ms depending on the method. (Eg. OVAS with size 8 chunks has delays in the range of 41- 62 ms with an effective delay of 70 - 90 ms) 

3. Part of the range of this delay is definitely due to Unity and the event recording as that code runs only at approx 250 Hz producing a +/- 5 ms range on when the hit event will be reported to LSL. 

Based on the above my plan was to record with whatever method has the least delay (most likely BrainVisionRDA.exe because I encountered some clipping issues with BrainAmpSeries.exe) and then remove the average delay from the EEG stream. That should be good enough for my analysis.

However if there was any way of dropping that recording delay or at least identifying the cause would be nice. I understand that a delay of about 50 - 100 ms is not outlandish but it does seem a bit odd as I would expect something in the region of an order of magnitude less. 

Again, many thank you very much for your help.

**Comment by ioannis-dz** [comment]
I run some more experiments, following a similar setup to the one for recording the BioSemi device latencies:

1.  I used a [Humisoft MF 624] (http://www.humusoft.cz/produkty/datacq/mf624/) DAQ card to produce a PWM pulse with a period of 250 ms and an amplitude of 50 mV
2. The PWM was then inserted to:
* a A/D input of the DAQ card
* a channel of the BrainAmp amplier.
3. Both streams were recorded by LSL and an Event marker was sent by the A/D recording program when the voltage was sensed to be higher than 30 mV.
4. The difference between the Event marker and the BrainAmp recordings is the delay caused by the Brain Amp. 

The results are an average delay of 51.6 ms ranging from 50.4 ms to 53 ms. Following are the histograms of the delay of two tests I ran, a 3 min and a 6 min recording:

![vodka_norda_ch1_delays](https://cloud.githubusercontent.com/assets/12684882/9192694/331bf1b6-4005-11e5-8d76-cb2c8ab756e6.png)

![vodka_norda_ch1_delays6mins](https://cloud.githubusercontent.com/assets/12684882/9192697/36f2e696-4005-11e5-9b00-909ef21174da.png)


The good news is that the delays are within a range of 2.6 ms so by removing a known delay of 51.7 we can get accuracies of a bit more than 1 ms which is great.

However, the delays are exceptionally high, about an order of magnitude higher than the delays of BioSemi in comparison. Especially for use with BCI such delays are very problematic.
